software/contrib/euclid.py

```diff
@@ -0,0 +1,482 @@
+        oled.text(f"-- {self.menu_items[menu_item]} --", 0, 0)
+        oled.text(f"{current_setting} <- {new_setting}", 0, 10)
+        oled.show()
+        
+    def apply_setting(self):
+        """Apply the current setting
+        """
+        (menu_item, lower_bound, upper_bound, current_setting, new_setting) = self.read_knobs()
+        
+        if menu_item == self.MENU_ITEMS_STEPS:
+            self.generator.steps = new_setting
+            if self.generator.pulses > new_setting:
+                self.generator.pulses = new_setting
+            if self.generator.rotation > new_setting:
+                self.generator.rotation = new_setting
+        elif menu_item == self.MENU_ITEMS_PULSES:
+            self.generator.pulses = new_setting
+        elif menu_item == self.MENU_ITEMS_ROTATION:
+            self.generator.rotation = new_setting
+        elif menu_item == self.MENU_ITEMS_SKIP:
+            self.generator.skip = new_setting / 100.0
+            
+        self.generator.regenerate()
+        
+
+class Screensaver:
+    """Blanks the screen to prevent burn-in
+    """
+    def __init__(self):
+        pass
+    
+    def draw(self):
+        oled.fill(0);
+        oled.show()
+
+class EuclideanRhythms(EuroPiScript):
+    """Generates 6 different Euclidean rhythms, one per output
+
+    Must be clocked externally into DIN
+    """
+    
+    def __init__(self):
+        super().__init__()
+        
+        ## The euclidean pattern generators for each CV output
+        #
+        #  We pre-load the defaults with some interesting patterns so the script
+        #  does _something_ out of the box
+        self.generators = [
+            EuclidGenerator(cv1, 8, 5),
+            EuclidGenerator(cv2, 16, 7),
+            EuclidGenerator(cv3, 16, 11),
+            EuclidGenerator(cv4, 32, 9),
+            EuclidGenerator(cv5, 32, 15),
+            EuclidGenerator(cv6, 32, 19)
+        ]
+        
+        self.load()
+        
+        self.channel_menu = ChannelMenu(self)
+        self.settings_menu = SettingsMenu(self)
+        self.screensaver = Screensaver()
+        
+        self.active_screen = self.channel_menu
+        
+        self.last_interaction_time = time.ticks_ms()
+        
+        @din.handler
+        def on_rising_clock():
+            """Handler for the rising edge of the input clock
+
+            Advance all of the rhythms
+            """
+            for g in self.generators:
+                g.advance()
+                
+        @din.handler_falling
+        def on_falling_clock():
+            """Handler for the falling edge of the input clock
+
+            Turn off all of the CVs so we don't stay on for adjacent pulses
+            """
+            for cv in cvs:
+                cv.off()
+            
+        @b1.handler
+        def on_b1_press():
+            """Handler for pressing button 1
+            """            
+            self.last_interaction_time = time.ticks_ms()
```

Comment:
[optional] I missed this the first time, but the EuroPi API already tracks button last press time: https://allen-synthesis.github.io/EuroPi/generated/europi.html#europi.Button.last_pressed

Feel free to submit a follow up patch to clean up the code, but not necessary.